gflownet/envs/tree.py

```diff
@@ -0,0 +1,614 @@
+
+        assert self.state[0] == Stage.THRESHOLD
+        assert attributes[0] == NodeType.CONDITION
+        assert torch.all(attributes[1:3] >= 0)
+        assert attributes[3] == -1
+        assert attributes[4] == Status.ACTIVE
+
+        attributes[3] = operator
+
+        self.state[0] = Stage.OPERATOR
+
+        self._split_leaf(k)
+
+    def _split_leaf(self, k: int) -> None:
+        """
+        Finalize the splitting of a leaf using set feature, threshold
+        and operator: transform the leaf into a condition node and
+        spawn two children.
+
+        We assume the convention of having left child output label = 0
+        and right child label = 1 if operator is <, and the opposite if
+        operator is >=. That way, during prediction, we can act as if
+        the operator was always the same (and only care about the output
+        label).
+        """
+        attributes = self._get_attributes(k)
+
+        assert self.state[0] == Stage.OPERATOR
+        assert attributes[0] == NodeType.CONDITION
+        assert torch.all(attributes[1:4] >= 0)
+        assert attributes[4] == Status.ACTIVE
+
+        k_left = Tree._get_left_child(k)
+        k_right = Tree._get_right_child(k)
+
+        if attributes[3] == Operator.LT:
+            self._insert_classifier(k_left, output=0)
+            self._insert_classifier(k_right, output=1)
+        else:
+            self._insert_classifier(k_left, output=1)
+            self._insert_classifier(k_right, output=0)
+
+        attributes[3] = -1
+        attributes[4] = Status.INACTIVE
+
+        self.leafs.remove(k)
+        self.state[0] = Stage.COMPLETE
+
+    def _insert_classifier(self, k: int, output: int) -> None:
+        """
+        Replace attributes of k-th node with those of a classifier node.
+        """
+        attributes = self._get_attributes(k)
+
+        assert torch.all(torch.isnan(attributes))
+
+        attributes[0] = NodeType.CLASSIFIER
+        attributes[1:3] = -1
+        attributes[3] = output
+        attributes[4] = Status.INACTIVE
+
+        self.leafs.add(k)
+
+    def get_action_space(self) -> List[Tuple[int, int]]:
+        """
+        Actions are a tuple containing:
+            1) action type:
+                0 - pick leaf to split,
+                1 - pick feature,
+                2 - pick threshold,
+                3 - pick operator,
+            2) node index.
+        """
+        actions = [(t, k) for t in range(4) for k in range(self.n_nodes)]
+        actions.append(self.eos)
+
+        return actions
+
+    def step(
+        self, action: Tuple[int, int, float]
+    ) -> Tuple[List[int], Tuple[int, int, float], bool]:
+        # If action not found in action space raise an error.
+        if action[:2] not in self.action_space:
+            raise ValueError(
+                f"Tried to execute action {action} not present in action space."
+            )
+        else:
+            action_idx = self.action_space.index(action[:2])
```

Comment:
Why `action[:2]` instead of just `action`?

Reply:
Ok, got it by myself - actions are tuples of three things. I got confused by the dosctring in the action space method. I am updating it.